subsys/bluetooth/audio/has.c

```diff
@@ -328,6 +301,89 @@ BT_CONN_CB_DEFINE(conn_cb) = {
 	.disconnected = disconnected,
 	.security_changed = security_changed,
 };
+
+static void notify_work_reschedule(struct has_client *client, enum has_notify notify,
+				   k_timeout_t delay)
+{
+	int err;
+
+	atomic_set_bit(client->notify, notify);
+
+	if (client->conn == NULL) {
+		return;
+	}
+
+	err = k_work_reschedule(&client->notify_work, delay);
+	if (err < 0) {
+		LOG_ERR("Failed to reschedule notification work err %d", err);
+	}
+}
+
+static void notify_work_handler(struct k_work *work)
+{
+	struct has_client *client = CONTAINER_OF(work, struct has_client, notify_work);
+	int err;
+
+#if defined(CONFIG_BT_HAS_FEATURES_NOTIFIABLE)
+	if (atomic_test_and_clear_bit(client->notify, NOTIFY_FEATURES) &&
+	    bt_gatt_is_subscribed(client->conn, FEATURES_ATTR, BT_GATT_CCC_NOTIFY)) {
+		err = bt_gatt_notify(client->conn, FEATURES_ATTR, &has.features,
+				     sizeof(has.features));
+		if (err == -ENOMEM) {
+			notify_work_reschedule(client, NOTIFY_FEATURES,
+					       K_USEC(BT_AUDIO_NOTIFY_RETRY_DELAY_US));
+		} else if (err < 0) {
+			LOG_ERR("Notify features err %d", err);
+		}
+	}
+#endif /* CONFIG_BT_HAS_FEATURES_NOTIFIABLE */
+
+#if defined(CONFIG_BT_HAS_PRESET_SUPPORT)
+	if (atomic_test_and_clear_bit(client->notify, NOTIFY_READ_PRESET_RESPONSE)) {
+		err = read_preset_response(client);
+		if (err == -ENOMEM) {
+			notify_work_reschedule(client, NOTIFY_READ_PRESET_RESPONSE,
+					       K_USEC(BT_AUDIO_NOTIFY_RETRY_DELAY_US));
+		} else if (err < 0) {
+			LOG_ERR("Notify read preset response err %d", err);
+		}
+	} else if (atomic_test_and_clear_bit(client->notify, NOTIFY_CURRENT_PRESET_LIST)) {
+		err = list_current_presets(client);
+		if (err == -ENOMEM) {
+			notify_work_reschedule(client, NOTIFY_CURRENT_PRESET_LIST,
+					       K_USEC(BT_AUDIO_NOTIFY_RETRY_DELAY_US));
+		} else if (err < 0) {
+			LOG_ERR("Notify generic update all presets err %d", err);
+		}
+	}
+
+	if (atomic_test_and_clear_bit(client->notify, NOTIFY_ACTIVE_INDEX) &&
+	    bt_gatt_is_subscribed(client->conn, ACTIVE_PRESET_INDEX_ATTR, BT_GATT_CCC_NOTIFY)) {
+		err = bt_gatt_notify(client->conn, ACTIVE_PRESET_INDEX_ATTR,
+				     &has.active_index, sizeof(has.active_index));
+		if (err == -ENOMEM) {
+			notify_work_reschedule(client, NOTIFY_ACTIVE_INDEX,
+					       K_USEC(BT_AUDIO_NOTIFY_RETRY_DELAY_US));
+		} else if (err < 0) {
+			LOG_ERR("Notify active index err %d", err);
+		}
+	}
+#endif /* CONFIG_BT_HAS_PRESET_SUPPORT */
+}
+
+static void notify(struct has_client *client, enum has_notify notify)
+{
+	if (client != NULL) {
+		notify_work_reschedule(client, notify, K_NO_WAIT);
```

Comment:
When calling `k_work_reschedule` with `K_NO_WAIT`, that effectively dismissed the earlier scheduled work, and attempts to schedule it without waiting. That means that if we had scheduled a workitem with `K_USEC(BT_AUDIO_NOTIFY_RETRY_DELAY_US)`, the delay will be dismissed and the notification will be attempted with K_NO_WAIT. 

I don't think it's a problem, but is this intentional? Or would it more sense to check if `k_work_delayable_is_pending` before scheduling with `K_NO_WAIT`?